connaisseur/validators/cosign/cosign_validator.py

```diff
@@ -43,25 +41,25 @@ async def validate(
             "threshold",
             1 if trust_root != "*" or any(required) else len(self.trust_roots),
         )
-
-        self.vals = self.__get_pinned_keys(trust_root, required, threshold)
+        # vals is a validations dict for each required trust root containing validated digests and errors
+        vals = self.__get_pinned_keys(trust_root, required, threshold)
 
         # use concurrent.futures for now
         # tasks = [self.__validation_task(k, str(image)) for k in self.vals.keys()]
         # await asyncio.gather(*tasks)
 
         # prepare executor
-        num_workers = len(self.vals)
+        num_workers = len(vals)
         executor = ThreadPoolExecutor(num_workers)
         # prepare tasks
-        arguments = [(k, str(image)) for k in self.vals.keys()]
+        arguments = [(k, v.copy(), str(image)) for k, v in vals.items()]
         futures = [executor.submit(self.__validation_task, *arg) for arg in arguments]
-        # await results (output dropped as `self.vals` is updated within function)
+        # await results (output dropped as `vals` is updated within function)
         for future in futures:
-            future.result()
+            vals.update(future.result())
```

Comment:
Does then the last error message win out? Should we be a bit more careful there and combine them to give accurate feedback?